doc/nnvpn.1
.TH NNVPN 1 "OCTOBER 2023" Linux "User Manuals"                                  
.SH NAME                                                                     
nnvpn \- Nnvpn is a VPN implementation using OpenSSL and Linux TUN devices.
.SH SYNOPSIS                                                                 
.B  nnvpn [-f config_path] [-s]
   [-d level]
   [-h] 
.SH DESCRIPTION                                                              
.B nnvpn 
this program implements a basic VPN based on two Linux computers, using OpenSSL for the cryptographic layer. 

A configuration file using LUA syntax must be provided to configure server and client conection parameters.

The program requires  cap_net_admin+ep capability, root user is not allowed.

.SH DISCLAIMERS
At the moment, This program is intended experimental.

.SH OPTIONS                                                       
.IP -f configuration_path
Specifies the full path related to the configuration file.
.IP -s
Server mode. If this option is specified, the program acts as server and not as client (default mode).
.IP -d level
Specifies debugging lev el (0-2).
.IP -h
A short description of arpchatcpp command line syntax.
.SH CONFIGURATION
This section explains the LUA configuration file syntax:
.IP Addess section
it specifies as string the server listening IP address or the remote server address in case of client mode, example:
.B  address = "192.168.0.111"
.IP Port section
it specifies as number the server listening IP port or the remote server port in case of client mode, example:
.B  port = 8181
.IP TUN Addess section
it specifies as string the TUN interface IP address, example:
.B  tunaddress = "10.0.0.1"
.IP TUN Netmask section
it specifies as string the TUN interface netmask, example:
.B  tunmask = "255.255.255.0"
.IP Certificate section
it specifies the path of OpenSSL pem file, example:
.B  cert = "/home/test/.keys/server.pem"
.IP Key section
it specifies the path of OpenSSL key file, example:
.B  key = "/home/test/.keys/server.key"
.IP Device section
it specifies as string the name of TUN device, example:
.B  device = "tun222"
.IP Psize section
it specifies as number the max payload size, it must be a multiple of 1500
.B  psize = 3000
.IP Log section
it specifies as string name and path of the required log file, example:
.B  log = "/tmp/nnvpn.log"
.SH BUGS                                                                     
This program is experimental, massive changes are possible.
.SH AUTHOR                                                                   
Gabriele Bonacini <gabriele.bonacini@protonmail.com>
.SH "SEE ALSO"                                                               
 OpenSSL, TUN.


ChangeLog
tvpn (0.2.2) any; urgency=low 

 * Beta Test
 Initial Version.
 -- Gabriele Bonacini <gabriele.bonacini@protonmail.com> Tue, 17 Oct 2023 21:39:00 -0100 



src/nnvpn.cpp
// -----------------------------------------------------------------
// nnvpn - a trivial VPN using OpenSSL
// Copyright (C) 2023  Gabriele Bonacini
//
// This program is distributed under dual license:
// - Creative Commons Attribution-NonCommercial 4.0 International (CC BY-NC 4.0) License 
// for non commercial use, the license has the following terms:
// * Attribution — You must give appropriate credit, provide a link to the license, 
// and indicate if changes were made. You may do so in any reasonable manner, 
// but not in any way that suggests the licensor endorses you or your use.
// * NonCommercial — You must not use the material for commercial purposes.
// A copy of the license it's available to the following address:
// http://creativecommons.org/licenses/by-nc/4.0/
// - For commercial use a specific license is available contacting the author.
// -----------------------------------------------------------------

#include <unistd.h>
#include <stdlib.h>

#include <string>
#include <iostream>

#include <parseCmdLine.hpp>
#include <debug.hpp>
#include <configFile.hpp>
#include <capabilities.hpp>
#include <inetgeneral.hpp>

using namespace std;
using namespace debugmode;
using namespace capabilities;
using namespace inetlib;

using parcmdline::ParseCmdLine;
using configFile::ConfigFileException;
using configFile::ConfigFile;
using configFile::IpAddr;

#ifdef __clang__
  void printInfo(char* cmd) __attribute__((noreturn));
#else
  [[ noreturn ]]
  void printInfo(char* cmd);
#endif

int main(int argc, char** argv){
    const long       MAX_PAYLOAD  { 1500 };
    const char       flags[]      { "hd:f:s"};
    DEBUG_MODE       debugMode    { DEBUG_MODE::ERR_DEBUG };
    string           configFile   { "./nnvpn.lua"};
    int              ret          { 0 };
    ParseCmdLine     pcl          {argc, argv, flags};
    bool             isServer     { false };

    if(pcl.getErrorState()){
        string exitMsg{string("Invalid  parameter or value").append(pcl.getErrorMsg())};
        cerr << exitMsg << "\n";
        printInfo(argv[0]);
    }

    if(pcl.isSet('h')) printInfo(argv[0]);
    if(pcl.isSet('s')) isServer = true;
    if(pcl.isSet('f')) configFile = pcl.getValue('f');

    if(pcl.isSet('d')){
            unsigned long debug{ stoul(pcl.getValue('d')) };
            switch(debug){
                case 0:
                    debugMode = DEBUG_MODE::ERR_DEBUG;
                   break;
                case 1: 
                    debugMode = DEBUG_MODE::STD_DEBUG;
                   break;
                case 2: 
                    debugMode = DEBUG_MODE::VERBOSE_DEBUG;
                   break;
                default:
                    debugMode = DEBUG_MODE::STD_DEBUG;
            }
            Debug::setDebugLevel(debugMode);
    }

    uint16_t         port         { 0 };
    long             psize        { MAX_PAYLOAD };
    string           address      { "" },
                     tunaddress   { "" },
                     tunmask      { "" },
                     cert         { "" },
                     key          { "" },
                     device       { "" },
                     logFile      { "" };
    try{
         ConfigFile cfg(configFile);
         try{
             cfg.init();
             cfg.addLoadableVariable("address", ""); 
             cfg.addLoadableVariable("port", 8081L, true);
             cfg.addLoadableVariable("psize", MAX_PAYLOAD, true);
             cfg.addLoadableVariable("cert", "");
             cfg.addLoadableVariable("key", "");
             cfg.addLoadableVariable("device", "");
             cfg.addLoadableVariable("log", "");
             cfg.addLoadableVariable("tunaddress", "");
             cfg.addLoadableVariable("tunmask", "");
    
             cfg.loadConfig();
    
             cfg.getConf("address").getIp(address);
             port     = cfg.getConf("port").getPort(); 
             psize    = cfg.getConf("psize").getInteger(); 
             if(psize < 0 || ( psize % MAX_PAYLOAD ) != 0 ) throw ConfigFileException("Invalid payload size");
             cert     = cfg.getConf("cert").getText();
             device   = cfg.getConf("device").getText();
             key      = cfg.getConf("key").getText();
             logFile  = cfg.getConf("log").getText();
             cfg.getConf("tunaddress").getIp(tunaddress);
             cfg.getConf("tunmask").getIp(tunmask);
         } catch(ConfigFileException& ex){
             ret = 1;
             string msg {"Error loading configuration file: "};
             msg.append(ex.what());
             cerr << msg << "\n";
             printInfo(argv[0]);
             throw string{"Abort."};
         }
  
         Debug debug{debugMode};
         try{
             debug.init(logFile);

         }catch(DebugException& ex){
             ret = 1;
             cerr << "Error: " << ex.what() << "\n";
             throw string{"Abort."};
         }
 
         Capability cpb;
         try{
             cpb.init(true); 
             cpb.reducePriv("cap_net_admin+ep");
             cpb.getCredential();
             if(debugMode > 1) cpb.printStatus();
         }catch(const CapabilityException& ex){
             ret = 2;
             cerr << "Error: " << ex.what() << "\n";
             throw string{"Abort."};
         }catch(...){
             ret = 2;
             cerr << "Error: unandled exception in privilege management." << "\n";
             throw string{"Abort."};
         }

         try{
             if(isServer){
                  NnVpnServer svpn(cert, key, address, to_string(port), device, psize);
                  svpn.init(tunaddress, tunmask);
                  svpn.start();
             } else {
                  NnVpnClient cvpn(cert, key, address, to_string(port), device, psize);;
                  cvpn.init(tunaddress, tunmask);
                  cvpn.start();
             }
         }catch(InetException& ex){
             ret = 3;
             cerr << "Error: " << ex.what() << "\n";
             throw string{"Abort."};
         }catch(...){ 
             ret = 1;
             throw string{"Unmanaged Error. Abort."};
         }

    }catch(const string& ex){
        cerr << ex << "\n";
        cout << "Program exits with error(s): check log file.\n";
    }

    return ret;  
}

void printInfo(char* cmd){
      cerr << cmd << " [-f <config_full_path>] [-d level] [-s] | [-h]\n\n";
      cerr << " -f  <full_path> Specify the configuration file path\n";
      cerr << " -d  <dbg_level> set debug mode\n";
      cerr << " -s              set server mode\n";
      cerr << " -h              print this synopsis\n";
      exit(EXIT_FAILURE);
}

